### ISSUE TITLE
Callbacks when network and account are changed

### ISSUE THREAD

**Issue by chevdor** [MetaMask/metamask-plugin#704]
I propose the addition of callbacks for the following cases:
1. network changed
2. address/account changed

That would allow DAPP to avoid polling for such critical information.

**Comment by kumavis** [comment]
you'll need to open this issue in [web3.js](https://github.com/ethereum/web3.js), we just follow that standard

**Comment by kumavis** [comment]
really this is best solved in user space. create a generic module for polling for this information. use that in your apps where its relevant to you.

**Comment by flyswatter** [comment]
He just did that, and they kicked him out:
https://github.com/ethereum/EIPs/issues/151#issuecomment-251701063

I think we've gotten a strong message to start adding features without regard for EIPs.

**Comment by flyswatter** [comment]
I think we could keep this issue open in any case, for "however we solve it".

**Comment by flyswatter** [comment]
These *could* be solved as pub-sub services on the node, but they're also things whose data source could easily be MetaMask itself, so maybe we just need to start iterating that feature, or maybe submit as a request to web3-js itself.

**Comment by kumavis** [comment]
this would be solved by `eth-store`

**Comment by chevdor** [comment]
The folks from web3 did not kick anything/anyone out 😄 , they actually patiently tried to understand what I was describing, thinking I was writing in this repo, which lead to a little confusion on both sides I think!

Sure a DAPP can poll but that makes no sense if we can get the event from another place where polling is not required.

Obviously, the extension knows  when the user switches network and/or account and could provide (that´s the request here) a callback.

**Comment by chevdor** [comment]
@kumavis would you have a link about +eth-store+ ?

Now the confusion makes somehow sense. 

I initially was thinking about Metamask only but we should keep Mist and any other upcoming option in mind. To me it sounds rather out of the scope of web3 (I may be wrong). Web3 could probably have a callback when the provider is set/changed but unless I am wrong, it is not aware of what is the current address the user wish to you. The address the user wish to use is specific to the client (Mist, Metamask, ...)

Initially, I was thinking about a way to register a few callback with Metamask to get feedback on what the user is doing in the extension (such as switching to another account).

If you see another way (besides polling 😄 ) to do that, please let me know.

**Comment by flyswatter** [comment]
Yeah I was being a little dramatic, but ultimately saying "this is out of scope" I think is very wrong there.

The only API surface MetaMask gives you, the developer, is the web3 object. So if we want to give you a new API surface, either we "get it approved" as a "standard", or we just tack it on.

I'm starting to accept that in a decentralized environment, there maybe is no approval or standards, and it may be a do-ocracy, so I'm looking at ways to expose a new API method, and *then* propose it as an addition to Web3.

**Comment by kumavis** [comment]
cool so we are going to build our own competing standard
expect this to be released around devcon3

( in the mean time, just create some reusable polling code. or stay tuned for [eth-store](https://github.com/kumavis/eth-store) )

**Comment by chevdor** [comment]
While this is not related to the 'account changed' topic, I am currently facing a similar issue that polling cannot solve:

I am trying to use Metamask for a DAPP. Here is the process as the user sees it:
- call a DAPP function
- confirm (or reject) the tx in the Metamask popup
- wait for tx
- wait for receipt
- get result

On a local node, getting a tx hash is almost instant. However, using Metamask and its node, it really takes a while for the DAPP to get back a tx. During that time, there is no way for the DAPP to tell the user what is currently happening. The bad case is when Metamask loses connection (I will open other issue about that). When that happens, the DAPP cannot know about it and waits forever. Sure there is a timeout after 4 minutes but why wait so long when we could tell the user right away.

Moreover, since the DAPP does not know when the user confirmed, it cannot provide feedback to the user. AFAIK, there is no way to poll Metamask to know if the user confirmed a tx.

I am neutral in the way the solution is implemented but I think a solution needs to be provided. As I see parity develops parity.js, there is also inject.js and of course web3.js

If web3 is incomplete, why not extend it or provide a Metamask specific object that could tell the DAPP a little more about Users actions in Metamask?

**Comment by kumavis** [comment]
@chevdor
> However, using Metamask and its node, it really takes a while for the DAPP to get back a tx

where are you geo-located? what does your rpc latency look like?
we can return the txHash slightly sooner but we need to do a few network lookups (nonce, gasPrice, gasLimit) before we can sign the tx and derive its hash. Zero-client will always be a slower than running a local node, but you shouldnt be seeing wait times >30 sec. I need to know your average rpc response time and tx signature time to better diagnose. are you on a resource limited device?

**Comment by kumavis** [comment]
you are welcome to open an issue to discuss long timeouts for signing and submitting txs, but I'm closing this issue for non-standard web3 methods.

**Comment by chevdor** [comment]
long tx issue opened as #817
